tests/e2e/tidbcluster/tidbcluster.go

```diff
@@ -1607,6 +1578,327 @@ var _ = ginkgo.Describe("TiDBCluster", func() {
+								wrongImage = false
+								break
+							}
+						}
+						if wrongImage {
+							log.Failf("%s Pod has wrong image, expected %s", comp, utilimage.TiDBV4)
+						}
+					}
+				})
+			}
+		})
+
+		ginkgo.It("PD to 0 is forbidden while other components are running", func() {
+			ginkgo.By("Deploy initial tc")
+			tc := fixture.GetTidbCluster(ns, "scale-pd-to-0", utilimage.TiDBV4)
+			tc.Spec.PD.Replicas = 1
+			utiltc.MustCreateTCWithComponentsReady(genericCli, oa, tc, 5*time.Minute, 10*time.Second)
+
+			ginkgo.By("Scale in PD to 0 replicas")
+			err := controller.GuaranteedUpdate(genericCli, tc, func() error {
+				tc.Spec.PD.Replicas = 0
+				return nil
+			})
+			framework.ExpectNoError(err, "failed to scale in PD for TidbCluster %s/%s", ns, tc.Name)
+
+			ginkgo.By("Wait for PD to be in ScalePhase")
+			utiltc.MustWaitForComponentPhase(cli, tc, v1alpha1.PDMemberType, v1alpha1.ScalePhase, 3*time.Minute, 10*time.Second)
+
+			ginkgo.By("Check for FailedScaleIn event")
+			// LAST SEEN   TYPE      REASON          OBJECT              MESSAGE
+			// 25s         Warning   FailedScaleIn   tidbcluster/basic   The PD is in use by TidbCluster [pingcap/basic], can't scale in PD, podname basic-pd-0
+			err = wait.Poll(5*time.Second, 1*time.Minute, func() (done bool, err error) {
+				options := metav1.ListOptions{
+					FieldSelector: fmt.Sprintf("involvedObject.name=%s,involvedObject.kind=TidbCluster", tc.Name),
+				}
+				event, err := c.CoreV1().Events(ns).List(options)
+				if err != nil {
+					log.Logf("failed to list events with options: +%v", options)
+					return false, nil
+				}
+				for _, event := range event.Items {
+					log.Logf("found event: %+v", event)
+					if event.Reason == "FailedScaleIn" && strings.Contains(event.Message, "PD") {
+						return true, nil
+					}
+				}
+				return false, nil
+			})
+			framework.ExpectNoError(err, "failed to wait for FailedScaleIn event")
+		})
+
+		ginkgo.It("TiKV from >=3 replicas to <3 should be forbidden", func() {
+			ginkgo.By("Deploy initial tc")
+			tc := fixture.GetTidbCluster(ns, "scale-in-tikv", utilimage.TiDBV4)
+			tc, err := cli.PingcapV1alpha1().TidbClusters(tc.Namespace).Create(tc)
+			framework.ExpectNoError(err, "Expected create tidbcluster")
+			err = oa.WaitForTidbClusterReady(tc, 30*time.Minute, 5*time.Second)
+			framework.ExpectNoError(err, "Expected get tidbcluster")
+
+			ginkgo.By("Scale in tikv to 2 replicas")
+			err = controller.GuaranteedUpdate(genericCli, tc, func() error {
+				tc.Spec.TiKV.Replicas = 2
+				return nil
+			})
+			framework.ExpectNoError(err, "failed to scale in tikv")
+
+			ginkgo.By("Expect up stores number stays 3")
+			pdClient, cancel, err := proxiedpdclient.NewProxiedPDClient(c, fw, ns, tc.Name, false)
+			framework.ExpectNoError(err, "create pdClient error")
+			defer cancel()
+			storesInfo, err := pdClient.GetStores()
+			framework.ExpectNoError(err, "get stores info error")
```

Comment:
Should we move these into the `wait` block below?

Reply:
fixed